plugins/inputs/syslog/syslog.go

```
@@ -0,0 +1,419 @@
+	s.connectionsMu.Lock()
+	delete(s.connections, c.RemoteAddr().String())
+	s.connectionsMu.Unlock()
+}
+
+func (s *Syslog) handle(conn net.Conn, acc telegraf.Accumulator) {
+	defer func() {
+		s.removeConnection(conn)
+		conn.Close()
+	}()
+
+	if s.ReadTimeout != nil && s.ReadTimeout.Duration > 0 {
+		conn.SetReadDeadline(time.Now().Add(s.ReadTimeout.Duration))
+	}
+
+	var p *rfc5425.Parser
+	if s.BestEffort {
+		p = rfc5425.NewParser(conn, rfc5425.WithBestEffort())
+	} else {
+		p = rfc5425.NewParser(conn)
+	}
+
+	p.ParseExecuting(func(r *rfc5425.Result) {
+		s.store(*r, acc)
+	})
+}
+
+func (s *Syslog) setKeepAlive(c *net.TCPConn) error {
+	if s.KeepAlivePeriod == nil {
+		return nil
+	}
+
+	if s.KeepAlivePeriod.Duration == 0 {
+		return c.SetKeepAlive(false)
+	}
+	if err := c.SetKeepAlive(true); err != nil {
+		return err
+	}
+	return c.SetKeepAlivePeriod(s.KeepAlivePeriod.Duration)
+}
+
+func (s *Syslog) store(res rfc5425.Result, acc telegraf.Accumulator) {
+	if res.Error != nil {
+		acc.AddError(res.Error)
+	}
+	if res.MessageError != nil {
+		acc.AddError(res.MessageError)
+	}
+	if res.Message != nil {
+		msg := *res.Message
+		acc.AddFields("syslog", fields(msg, s), tags(msg), s.time())
+	}
+}
+
+func tags(msg rfc5424.SyslogMessage) map[string]string {
+	ts := map[string]string{}
+
+	// Not checking assuming a minimally valid message
+	ts["severity"] = *msg.SeverityShortLevel()
+	ts["facility"] = *msg.FacilityLevel()
+
+	if msg.Hostname() != nil {
+		ts["hostname"] = *msg.Hostname()
+	}
+
+	if msg.Appname() != nil {
+		ts["appname"] = *msg.Appname()
+	}
+
+	return ts
+}
+
+func fields(msg rfc5424.SyslogMessage, s *Syslog) map[string]interface{} {
+	// Not checking assuming a minimally valid message
+	flds := map[string]interface{}{
+		"version": msg.Version(),
+	}
+	flds["severity_code"] = int(*msg.Severity())
+	flds["facility_code"] = int(*msg.Facility())
+
+	if msg.Timestamp() != nil {
+		flds["timestamp"] = *msg.Timestamp()
```

I think this doesn't work because this is a time.Time and a field has to be one of these types.  If this returns nil then the field is silently dropped!

https://github.com/influxdata/telegraf/blob/ce3b367dacd415d5b9fbe3be104e671fe51546af/metric/metric.go#L251-L286

---

Probably should encode as an integer in nanosecond precision

---

:+1: